client/components/date-range/test/index.js

```diff
@@ -0,0 +1,632 @@
+			const callback = jest.fn();
+
+			const wrapper = shallow( <DateRange moment={ moment } onDateCommit={ callback } /> );
+
+			const newStartDate = moment( '2018-04-01' );
+			const newEndDate = moment( '2018-04-29' );
+
+			// Select dates using API
+			wrapper.instance().onSelectDate( newStartDate );
+			wrapper.instance().onSelectDate( newEndDate );
+
+			// Commit/apply those dates
+			wrapper.instance().commitDates();
+
+			expect( callback ).toHaveBeenCalledTimes( 1 );
+			expect( dateToLocalString( callback.mock.calls[ 0 ][ 0 ] ) ).toEqual(
+				dateToLocalString( newStartDate )
+			);
+			expect( dateToLocalString( callback.mock.calls[ 0 ][ 1 ] ) ).toEqual(
+				dateToLocalString( newEndDate )
+			);
+		} );
+	} );
+
+	describe( 'Apply and cancel', () => {
+		test( 'should only persist date selection when user clicks "Apply" button', () => {
+			const wrapper = shallow( <DateRange moment={ moment } /> );
+			const originalStartDate = wrapper.state().startDate;
+			const originalEndDate = wrapper.state().endDate;
+
+			// Get child components
+			const dateRangeHeader = wrapper.find( DateRangeHeader );
+			const datePicker = wrapper.find( DatePicker );
+
+			const newStartDate = moment( '2018-04-01' );
+			const newEndDate = moment( '2018-04-29' );
+
+			// Select dates using API
+			datePicker.props().onSelectDay( newStartDate );
+			datePicker.props().onSelectDay( newEndDate );
+
+			// Force update
+			wrapper.update();
+
+			// Calls whichever method DateRange passes into DateRangeHeader component
+			dateRangeHeader.props().onCancelClick();
+
+			// Should still be the original dates...
+			expect( wrapper.state() ).toEqual(
+				expect.objectContaining( {
+					startDate: originalStartDate,
+					endDate: originalEndDate,
+				} )
+			);
+
+			// Select dates using API
+			datePicker.props().onSelectDay( newStartDate );
+			datePicker.props().onSelectDay( newEndDate );
+
+			// Force update
+			wrapper.update();
+
+			// Calls whichever method DateRange passes into DateRangeHeader component
+			dateRangeHeader.props().onApplyClick();
+
+			// Should now be persisted
+			expect( dateToLocalString( wrapper.state().startDate ) ).toEqual(
+				dateToLocalString( newStartDate )
+			);
+			expect( dateToLocalString( wrapper.state().endDate ) ).toEqual(
+				dateToLocalString( newEndDate )
+			);
+		} );
+	} );
+
+	describe( 'Render props', () => {
+		test( 'should allow for render prop to overide trigger render', () => {
+			const spyComponent = jest.fn();
+
+			shallow( <DateRange moment={ moment } renderTrigger={ spyComponent } /> );
+
+			const props = spyComponent.mock.calls[ 0 ][ 0 ];
+			const propKeys = Object.keys( props ).sort();
+
+			expect( spyComponent ).toHaveBeenCalledTimes( 1 );
+
+			expect( propKeys ).toEqual(
+				[
+					'startDateText',
+					'endDateText',
+					'isCompact',
+					'buttonRef',
+					'onTriggerClick',
+					'triggerText',
+				].sort()
```

Comment:
Could we not just define these keys ready-sorted?

Reply:
Actually, I see the value of just allowing `[].sort()` to handle the sorting as it's the same `sort`ing as in done in `Object.keys( props ).sort();` above 👍

Reply:
Was trying to avoid a situation whereby order the order of props in the component would cause tests to fail.